src/nvim/api/win_config.c

```diff
@@ -386,7 +387,7 @@ static void parse_border_style(Object style, FloatConfig *fconfig, Error *err)
         return;
       }
       if (string.size
-          && mb_string2cells_len((char_u *)string.data, string.size) > 1) {
+          && vim_strnsize((char_u *)string.data, (int)string.size) > 1) {
```

Comment:
hmm, `vim_strnsize` needs to be fixed to use `utfc_ptr2len_len`. after that we could eliminate mb_string2cells_len. It is only used in one other place and deleting a redundant but buggy function would be good.

Reply:
That defeats the point of this PR. `utfc_ptr2len_len` treats control characters as 1-wide. I think it is intended for all functions in `mbyte.c` to treat control characters as 1-wide.

Reply:
BTW I also created <https://github.com/vim/vim/pull/9369> at the same when I created this PR.

Reply:
@zeertzjq what point? utfc_ptr2len_len is just the same as utfc_ptr2len except that it actually uses the width. sending a ptr,size pair to `vim_strnsize` is completely pointless if size is ignored and it still reads past the end of `ptr+size`. the *width* calculation itself should not change, which I never suggested.

Reply:
The point of this PR is to disallow control characters as border by detecting them as multi-width, like <https://github.com/vim/vim/pull/9369>.

Reply:
Wait a minute

Reply:
that's fine, but if there is a memory correctness issue, why not fix it at the same time?

Reply:
By `wide` I mean cells here. `vim_strnsize` doesn't check for byte length of a string. It checks for how many cells it occupies, and it detects control characters as multi-cells by calling `ptr2cells` instead of `utf_ptr2cells`.

Reply:
This name `vim_strnsize` is indeed misleading.

Reply:
> memory correctness issue

What?

Reply:
`vim_strnsize` takes a ptr/size pair. this implies it shouldn't access memory beyond `ptr+size`. this is trivial to fix by using utfc_ptr2len_len. This doesn't affect width calculation in any way. with this change `mb_string2cells_len` can be eliminated and `vim_strnsize`. In addition, the latter can be renamed if we so want, but that's secondary to actually fixing the issue.

Reply:
Oh, `vim_strnsize` indeed should be used in `gM` instead of `mb_string2cells_len`. This is also applicable to Vim. I'll make a Vim patch.

Reply:
For `gM` something other than `mb_string2cells_len` and `vim_strnsize` should be needed, as it needs to take Tab width into consideration. I've reported this it Vim (<https://github.com/vim/vim/issues/9409>). `vim_strnsize` may be needed for some virtual text related stuff.

Reply:
Oh, I finally understood what you meant.